datajoint/external.py

```diff
@@ -6,6 +6,203 @@
+        super().__init__(store, database)
+        self.check_required(store, 's3', RawFileHandler.required)
+        self._location = self._spec['location']
+
+    def get_folder(self):
+        return os.path.join(self._location, self._database)
+
+    def put(self, obj):
+        (blob, hash) = self.hash_obj(obj)
+
+        folder = self.get_folder()
+        full_path = os.path.join(folder, hash)
+        if not os.path.isfile(full_path):
+            try:
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+            except FileNotFoundError:
+                os.makedirs(folder)
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+
+        return (blob, hash)
+
+    def get_blob(self, hash):
+        full_path = os.path.join(self.get_folder(), hash)
+        try:
+            with open(full_path, 'rb') as f:
+                return f.read()
+        except FileNotFoundError:
+                raise DataJointError('Lost external blob')
+
+    def get_obj(self, hash):
+        return unpack(self.get_blob(hash))
+
+    def get(self, hash):
+        return self.get_obj(hash)
+
+
+class CacheFileHandler(ExternalFileHandler):
+    '''
+    A CacheFileHandler mixin implementation.
+
+    Requires a concrete 'upstream' backend file handling implementation.
+    Should be 1st in inheritance list w/r/t backend implementation - e.g.:
+
+      CachedFoo(CacheFileHandler, OtherFileHandler)
+
+    Will cache objects in config['external-cache']['location'], relying on
+    superclass for coherent get/put operations on the 'reference' blobs.
+
+    Cleanup currently not implemented.
+    '''
+
+    def __init__(self, store, database):
+        super().__init__(store, database)  # initialize non-mixin parameters
+
+        # validate mixin cache parameters
+        if 'external-cache' not in config:
+            raise DataJointError('External Cache is not configured')
+
+        cache_spec = config['external-cache']
+
+        if 'location' not in cache_spec:
+            raise DataJointError(
+                'External Cache configuration missing "location"')
+
+        self._cache_spec = cache_spec
+        self._cache_location = cache_spec['location']
+
+    def get_cache_folder(self):
+        return os.path.join(self._cache_location, self._database)
+
+    def _clean_cache(self):
+        pass  # TODO: implement _clean_cache
+
+    def _put_cache(self, blob, hash):
+
+        self._clean_cache()
+
+        folder = self.get_cache_folder()
+        full_path = os.path.join(folder, hash)
+
+        if not os.path.isfile(full_path):
+            try:
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+            except FileNotFoundError:
+                os.makedirs(folder)
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+        else:
+            pass  # TODO: track recent file usage to assist _clean_cache
```

Comment:
Incomplete TODO.

Let's not leave TODOs in code unless it's work in progress not ready for a pull request.  If you plan to implement a feature, propose it in issues where we can discuss it.

Reply:
I agree generally -
However, here, planning ahead for known-immediately-pending features is useful to prevent poor structure which will force subsequent rewriting; annotating and saving this effort is similarly useful.